multigeiger/transmission.cpp

```diff
@@ -0,0 +1,202 @@
+// measurements data transmission related code
+// - via WiFi to internet servers
+// - via LoRa to TTN (to internet servers)
+
+#include <Arduino.h>
+#include <HTTPClient.h>
+
+#include "log.h"
+#include "display.h"
+#include "userdefines.h"
+
+// Check if a CPU (board) with LoRa is selected. If not, deactivate SEND2LORA.
+#if !((CPU==LORA) || (CPU==STICK))
+#undef SEND2LORA
+#define SEND2LORA 0
+#endif
+
+#if SEND2LORA==1
+#include "loraWan.h"
+#endif
+
+#include "transmission.h"
+
+// Hosts for data delivery
+#define MADAVI "http://api-rrd.madavi.de/data.php"
+#define SENSORCOMMUNITY "http://api.sensor.community/v1/push-sensor-data/"
+#define TOILET "http://ptsv2.com/t/enbwck3/post"
+
+static String http_software_version;
+#if SEND2LORA
+static unsigned int lora_software_version;
+#endif
+static String chipID;
+
+void setup_transmission(const char *version, char *ssid) {
+  chipID = String(ssid);
+  chipID.replace("ESP32", "esp32");
+
+  http_software_version = String(version);
+
+  #if SEND2LORA
+  int major, minor, patch;
+  sscanf(version, "V%d.%d.%d", &major, &minor, &patch);
+  lora_software_version = (major << 12) + (minor << 4) + patch;
+  setup_lorawan();
+  #endif
+}
+
+void prepare_http(HTTPClient *http, const char *host) {
+  http->begin(host);
+  http->addHeader("Content-Type", "application/json; charset=UTF-8");
+  http->addHeader("Connection", "close");
+  http->addHeader("X-Sensor", chipID);
+}
+
+void send_http(HTTPClient *http, String body) {
+  if (DEBUG_SERVER_SEND == 1)
+    log(DEBUG, "http request body: %s", body.c_str());
+
+  int httpResponseCode = http->POST(body);
+  if (httpResponseCode > 0) {
+    String response = http->getString();
+    if (DEBUG_SERVER_SEND == 1) {
+      log(DEBUG, "http code: %d", httpResponseCode);
+      log(DEBUG, "http response: %s", response.c_str());
+    }
+  } else {
+    log(ERROR, "Error on sending POST: %d", httpResponseCode);
+  }
+  http->end();
+}
+
+void send_http_geiger(const char *host, String tube_type, unsigned int timediff, unsigned int hv_pulses, unsigned int gm_counts, unsigned int cpm) {
+  bool addname = false;
+  char body[1000];
+  HTTPClient http;
+  prepare_http(&http, host);
+  http.addHeader("X-PIN", "19");
+  tube_type = tube_type.substring(10);
+  String prefix = (addname ? tube_type + "_" : "");
+  const char *json_format = R"=====(
+{
+ "software_version": "%s",
+ "sensordatavalues": [
+  {"value_type": "%scounts_per_minute", "value": "%d"},
+  {"value_type": "hv_pulses", "value": "%d"},
+  {"value_type": "counts", "value": "%d"},
+  {"value_type": "sample_time_ms", "value": "%d"}
+ ]
+}
+)=====";
+  snprintf(body, 1000, json_format,
+           http_software_version,
```

Comment:
Either add here .c_str()  (-> http_software_version.c_str() ) or declare http_software_version as char* .
The %s cannot be used with String !
A few lines down it comes again.

Reply:
right, fixing...

Reply:
fixed.